File: test_ipc_mongo.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

import sys
import binascii
import threading
import time
import argparse
import subprocess
import signal
from bson.binary import Binary

import ipc.Ipc as Ipc
from ExampleProtocol.ExampleProtocol import ExampleMessage
import ipc.MongoIpc as MongoIpc
from defs import *

usleep = lambda x: time.sleep(x/1000000.0)

def signal_handler(signal, frame):
	subprocess.call(['mongod', '-f', '/home/routeflow/test_ipc_mongo/mongo/conf/master.conf', '--replSet', 'rs0', '--shutdown'])
	subprocess.call(['mongod', '-f', '/home/routeflow/test_ipc_mongo/mongo/conf/slave1.conf', '--replSet', 'rs0', '--shutdown'])
	subprocess.call(['mongod', '-f', '/home/routeflow/test_ipc_mongo/mongo/conf/slave2.conf', '--replSet', 'rs0', '--shutdown'])
	subprocess.call(['rm', '-rf', '/home/routeflow/test_ipc_mongo/mongo/data/*/*'])
	sys.exit(0)

class ExampleProtocolProcessor(Ipc.IpcMessageProcessor):
    def __init__(self):
	print ''

    def process(self, msg):
        type_ = msg.get_type()
        if type_ == EXAMPLE_MESSAGE:
            print msg.str()

if __name__ == "__main__":

	signal.signal(signal.SIGINT, signal_handler)
	description='Test for MongoIPC(Python) '
	epilog=''
	parser = argparse.ArgumentParser(description=description, epilog=epilog)
	parser.add_argument('sender', help='Sender of messages')
	parser.add_argument('mongo_collection', help='Mongo Collection')
	parser.add_argument('to', help='Send messages to')
	args = parser.parse_args()
	print 'Starting Test...'
	subprocess.call(['mongod', '-f', '/home/routeflow/test_ipc_mongo/mongo/conf/master.conf', '--replSet', 'rs0', '--shutdown'])
	subprocess.call(['mongod', '-f', '/home/routeflow/test_ipc_mongo/mongo/conf/slave1.conf', '--replSet', 'rs0', '--shutdown'])
	subprocess.call(['mongod', '-f', '/home/routeflow/test_ipc_mongo/mongo/conf/slave2.conf', '--replSet', 'rs0', '--shutdown'])
	subprocess.call(['rm', '-rf', '/home/routeflow/test_ipc_mongo/mongo/data/*/*'])
	subprocess.call('./init_mongo_replication.sh')
	
	# The second argument must specify the mongo collection 
	ipc = MongoIpc.MongoIpc(args.sender, args.mongo_collection)    
	# A processor is responsible for doing whatever your program wants to do
	# Waiting the messages in separate thread
	ipc.parallel_listen(ExampleProtocolProcessor())
	    
	# Define qtd messages 
	msgQtde = 1000

	print 'Sending ', msgQtde, ' messages'
	cal = (msgQtde/2.0) + 1.0
	while msgQtde > 0:
		# Time between each message
		usleep(100)

		# changing last message
		if msgQtde == 1:
		    data="last_message"
		else:
		    data="test_replication"
		
		# We create our message object (which is based on IPCMessage)
		m = ExampleMessage(True, 1, data)

		#Configuring sender and destination
		m.set_from(args.sender)
		m.set_to(args.to)

		# And then we send it using the service
		ipc.send(m)

		# stopped condition (mensagens/2)
		if msgQtde == cal:
			subprocess.call(['mongod', '-f', '/home/routeflow/test_ipc_mongo/mongo/conf/master.conf', '--replSet', 'rs0', '--shutdown'])
			time.sleep(5)
		msgQtde = msgQtde - 1
	print 'end of test'
	sys.exit()


	   

File: test_ipc_mongo.cc
#include <Ipc.h>
#include <MongoIpc.h>
#include <RFProtocol.h>
#include <ExampleProtocol/ExampleProtocol.h>
#include <IpcMessageProcessor.h>
#include <IpcMessage.h>
#include <stdio.h>
#include <signal.h>

class ExampleProtocolProcessor : public IpcMessageProcessor {
    public:
        ExampleProtocolProcessor() { }

    private:

        virtual void process(IpcMessage* msg) {
            cout << endl;
            cout << "Message from " << msg->getFrom() << " to " << msg->getTo() << endl;
            int type = msg->get_type();
            if (type == EXAMPLE_MESSAGE) {
                // Remember to use a dynamic_cast
                ExampleMessage *ex = dynamic_cast<ExampleMessage*>(msg);
                // Our example processing prints the message
                cout << ex->str() << endl;
            }
            cout << "-> Destination: " << endl;
            //return true;
        }
};
void signal_callback_handler(int signum){
   printf("Caught signal %d\n",signum);
   // Cleanup and close up stuff here

    system("mongod -f /home/routeflow/test_ipc_mongo/mongo/conf/master.conf --replSet rs0 --shutdown");
    system("mongod -f /home/routeflow/test_ipc_mongo/mongo/conf/slave1.conf --replSet rs0 --shutdown");
    system("mongod -f /home/routeflow/test_ipc_mongo/mongo/conf/slave2.conf --replSet rs0 --shutdown");
    system("rm -rf /home/routeflow/test_ipc_mongo/mongo/data/*/*");

   // Terminate program
   exit(signum);
}

int main(int argc, char* argv[]) {

    signal(SIGINT, signal_callback_handler);

	if (argc < 3) {
		cout << "Usage:" << endl;
		cout << "  " << "./ipc [sender] [mongobd_collection] [destination]" << endl;
		exit(EXIT_FAILURE);
	}

    // script of Mongo Replication
    system("/home/routeflow/test_ipc_mongo/./init_mongo_replication.sh");

    // We need to establish the IPC service that will send and receive messages
    // The first argument must specify the sender
    // The second argument must specify the mongo collection 
    MongoIpc ipc(argv[1], argv[2]);
    
    // A processor is responsible for doing whatever your program wants to do
    // When it receives a message
    ExampleProtocolProcessor *processor = new ExampleProtocolProcessor();
    
    // Waiting the messages in separate thread
    ipc.parallelListen(processor);  
    
    // Define qtd messages 
    long int msgQtde = 1000000;

    double cal = (msgQtde/2.0) + 1.0;
    string destination, data;

    while (msgQtde > 0) {

        // Time between each message
        usleep(100);

        // changing last message
        if (msgQtde == 1)
            data="last_message"; 
        else
            data="test_replication"; 
        
        // We create our message object (which is based on IPCMessage)
        ExampleMessage m(true, 1, data);

        //Configuring sender and destination
        m.setFrom(argv[1]);
        m.setTo(argv[3]);

        // And then we send it using the service
        ipc.send(&m);

        // stopped condition (mensagens/2)
        if (msgQtde == cal) {
            system("mongod -f /home/routeflow/test_ipc_mongo/mongo/conf/master.conf --replSet rs0 --shutdown");
            sleep(5);
        }       
        msgQtde--;
    }

    while (true) { 
    sleep(25);
    cout << "\n\n--> Press Ctrl+c to finish this program correctly...\n\n\n";
    }

    getchar();
    return 0;
}
